design/one-pager-stack-ui-metadata.md

```diff
@@ -0,0 +1,155 @@
+
+The optional addition of a `ui-schema.yaml` file within the package tree will be used to drive the UI.
+
+The primary purpose of the `ui-schema.yaml` format is to allow for easy authoring of the definition of UI markup, validation, and errors for a more complete UI and UX. The `ui-schema.yaml` file will be parsed and validated as part of package validation and ultimately serialized as a `JSON` annotation on the respective CRD at install time.
+
+### Example addition of `ui-schema.yaml` in a package tree
+
+```
+.registry/
+├── icon.png
+├── app.yaml # Application metadata.
+├── ui-schema.yaml # Optional UI spec for configuration metadata
+├── install.yaml # Optional install metadata.
+├── rbac.yaml # Optional RBAC permissions.
+└── resources
+      └── databases.foocompany.io # Group directory
+            ├── group.yaml # Optional Group metadata
+            ├── icon.png # Optional Group icon
+            └── mysql # Kind directory by convention
+                └── v1alpha1
+                    ├── mysql.v1alpha1.crd.yaml # Required CRD
+                    ├── ui-schema.yaml # Optional UI spec for configuration metadata
+                    ├── icon.png # Optional resource icon
+                    ├── resource.yaml # Resource level metadata.
+```
+
+### Example ui-schema.yaml
+
+This file contains multiple sections and a variety of different input types with validation overrides, extended validation and custom error messages.
+
+```yaml
+uiSpecVersion: 0.3
+uiSpec:
+- title: Configuration
+  description: Enter information specific to the configuration you wish to create.
+  items:
+  - name: dbReplicas
+    controlType: singleInput
+    type: integer
+    path: .spec.dbReplicas
+    title: DB Replicas
+    description: The number of DB Replicas
+    default: 1
+    validation:
+    - minimum: 1
+    - maximum: 3
+  - name: masterPassword
+    controlType: singleInput
+    type: password
+    path: .spec.masterPassword
+    title: DB Master Password
+    description: The master DB password. Must be between 8-32 characters long
+  - name: subdomain
+    controlType: singleInput
+    type: string
+    path: .spec.subdomain
+    title: Subdomain
+    pattern: ^([A-Za-z0-9](?:(?:[-A-Za-z0-9]){0,61}[A-Za-z0-9])?){2,62}$
+    description: Enter a value for your subdomain. It cannot start or end with a dash and must be between 2-62 characters long
+    validation:
+    - minLength: 2
+    - maxLength: 62
+  - name: instanceSize
+    controlType: singleSelect
+    path: .spec.instanceSize
+    title: Instance Size
+    enum:
+    - 'Option-1'
+    - 'Option-2'
+    - 'Option-3'
+    validation:
+    - required: true
+      customError: You must select an instance size for your configuration!
+```
+
+### CRD Annotation Example
+
+An example injection of the `ui-schema.yaml` as a CRD annotation would present as follows:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
```

Comment:
is there a max length limit for annotations on objects in the k8s api?

Reply:
3 years ago it increased from 64k to 256k

https://github.com/kubernetes/kubernetes/blob/0b4275b6c7ecd55cfd9916dbef8fe48ae4a8c733/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L35

Reply:
Noted this in 13b4da9